### ISSUE TITLE
Optional Template for New File - Feature Request

### ISSUE THREAD

**Issue by NicholasLinderWood** [asottile/babi#360]
hello hello! Thank you for creating babi as well as for all of your other educational material! I've been using babi and was wondering about a possible new feature which would (optionally) populate the editor with template text when creating a new file. After looking through the source code, here is one idea I've come up with for how this might be implemented.

1. Add an optional command line argument indicating whether to use a template (and if so which. only applicable if no filename is given or if the given filename doesn't exist)
2. Pass that argument along to the [File.ensure_loaded](https://github.com/asottile/babi/blob/main/babi/main.py#L25) method.
3.  In the case where the passed filename does not exist or filename is None, instead of passing the empty string to [get_lines](https://github.com/asottile/babi/blob/main/babi/file.py#L303), call a function which returns a string to pass in based on that command line argument (and perhaps other info). 
4. Create `babi/resources/templates.py` which will:
  a. Store various text templates
  b.  Contain the function referenced in 3.

If we think there would reasonably be only one template per file extension, then I'd lean toward making the command line argument a flag and then parse the filename to determine which flag to use. Then we could have something like `babi t.py --template` yielding

![image](https://github.com/user-attachments/assets/1f7bbf83-4a96-43a5-87ee-aad964dad886)

If not, then I think the command line argument would need to indicate which template to use. 

I'd be very interested in hearing your thoughts. Thanks for your time and consideration!

**Comment by asottile** [comment]
I don't think a command line argument is a good idea as it's really only usable *outside* the editor (and only on empty files).  I was thinking more of a `<esc>` command like `:m macroname` and a way for the user to have snippets in `~/.config/babi/snippets/...` or something

**Comment by NicholasLinderWood** [comment]
Okay, I've finally had some time to think more about this (it's only been a year and a half. . . ).

I'm thinking this could be implemented in a way that somewhat mimics cut/uncut? The user could select a certain amount of text and then do an `<esc>` command to create the snippet, something like `:snip snippet_name` (instead of cutting the selection we'd just  write it to `~/.config/babi/snippets/snippet_name`). Then, the user could follow-up with the `<esc>` command `:m snippet_name` which would insert the contents of that snippet into the current file (where I imagine the implementation would be nearly identical to `uncut`, but it would source the text from the snippet instead of the `cut_buffer`).

Is this along the lines of what you had in mind? Below are some mock-up screen shots. I think the names of these commands could use some work to make it clearer how they are connected to one-another.

<img width="683" height="364" alt="Image" src="https://github.com/user-attachments/assets/51d27264-ec20-4df6-8861-ca4e27180163" />

<img width="683" height="364" alt="Image" src="https://github.com/user-attachments/assets/307c442c-cbd2-4996-81fb-861cde17d77a" />

<img width="683" height="364" alt="Image" src="https://github.com/user-attachments/assets/ad5f14f3-e5fd-487f-88cd-d9c677f8256a" />

<img width="683" height="364" alt="Image" src="https://github.com/user-attachments/assets/708241d6-5d32-4dd2-bbf3-e1ff226e8438" />

**Comment by asottile** [comment]
I think there's not really a reason to make a special way to save them -- they can just be regular files edited normally

**Comment by NicholasLinderWood** [comment]
Cool I think that would simplify things. It could take another year and a half, but mind if I take a crack at this?

**Comment by asottile** [comment]
yeah feel free!  tbh it might take me a year to review it too lol I've been quite busy myself :(

**Comment by NicholasLinderWood** [comment]
Cool, thanks! And I'm in not in a huge rush so that's fine. I appreciate the input and I'll see what I can do. Who knows, maybe in two years babi will have a new feature :)
